### PR TITLE
[18.0][MIG] mgmtsystem_nonconformity: Migration to 18.0

### PR DESCRIPTION
**Pull request by natuan9**
**Depend on:**
- [[18.0][MIG] document_page_procedure: Migration to 18.0](https://github.com/natuan9/management-system/pull/1)
- `[[18.0][MIG] mgmtsystem_action: Migration to 18.0 ](https://github.com/OCA/management-system/pull/605)`

**Ported from v14:**
- `#525`

**Change in v18:**
 - `[remove order parameter of group_expand methods](https://github.com/odoo/odoo/pull/139294/commits/6e6bcbc6ab5cef679b17b617d821b2bb3f778af8) `

### CODE REVIEW

File: mgmtsystem_nonconformity/models/mgmtsystem_nonconformity_abstract.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from lxml import etree
+
+from odoo import api, fields, models
+from odoo.tools.misc import frozendict
+
+
+class MgmtsystemNonconformityAbstract(models.AbstractModel):
+    # TODO: Remove this on 17.0 and move everything on mail.thread
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/models/mgmtsystem_nonconformity_abstract.py]: Get it done bro :wink:

File: mgmtsystem_nonconformity/views/mgmtsystem_nonconformity.xml [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+            </search>
+        </field>
+    </record>
+
+    <!-- Kanban view -->
+    <record id="view_mgmtsystem_nonconformity_kanban" model="ir.ui.view">
+        <field name="name">mgmtsystem.nonconformity.kanban</field>
+        <field name="model">mgmtsystem.nonconformity</field>
+        <field name="arch" type="xml">
+            <kanban
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/views/mgmtsystem_nonconformity.xml:124]: Edit the kanban view
image

File: mgmtsystem_nonconformity/views/mgmtsystem_nonconformity.xml [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+                                                type="edit"
+                                                role="menuitem"
+                                                class="dropdown-item"
+                                            >
+                                                Edit
+                                            </a>
+                                        </li>
+                                    </t>
+                                </ul>
+                            </div>
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/views/mgmtsystem_nonconformity.xml]: Do we still need it?

File: mgmtsystem_nonconformity/models/mgmtsystem_nonconformity_abstract.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    _name = "mgmtsystem.nonconformity.abstract"
+    _description = "Nonconformity Abstract"
+
+    non_conformity_ids = fields.One2many(
+        "mgmtsystem.nonconformity",
+        inverse_name="res_id",
+        domain=lambda r: [("res_model", "=", r._name)],
+        readonly=True,
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/models/mgmtsystem_nonconformity_abstract.py]: If a field is marked as readonly=True on the model, it won't be possible to import it through the Odoo import tool, so avoid it as possible and define the readonly attribute on the views instead.
```suggestion
```

File: mgmtsystem_nonconformity/models/mgmtsystem_nonconformity.py [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    days_since_updated = fields.Integer(
+        readonly=True, compute="_compute_days_since_updated", store=True
+    )
+    number_of_days_to_close = fields.Integer(
+        "# of days to close",
+        compute="_compute_number_of_days_to_close",
+        store=True,
+        readonly=True,
+    )
+    closing_date = fields.Datetime(readonly=True)
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/models/mgmtsystem_nonconformity.py]: Same here

File: mgmtsystem_nonconformity/models/mgmtsystem_nonconformity.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+        "# of nonconformities", readonly=True, default=1
+    )
+    days_since_updated = fields.Integer(
+        readonly=True, compute="_compute_days_since_updated", store=True
+    )
+    number_of_days_to_close = fields.Integer(
+        "# of days to close",
+        compute="_compute_number_of_days_to_close",
+        store=True,
+        readonly=True,
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/models/mgmtsystem_nonconformity.py]: Same here. Let's fix it all :wink:

**Review by thienvh332 — CHANGES_REQUESTED** [review]

**Comment by natuan9** [comment]
Hi a @thienvh332 , Please help to review PR again

**Kanban Migration**
- Version 17.0
image

- Version 18.0
image

File: mgmtsystem_nonconformity/models/mgmtsystem_nonconformity.py [hunk 7]
```diff
@@ -19,27 +19,24 @@ def _default_stage(self):
         ].search([("is_starting", "=", True)], limit=1)
 
     @api.model
-    def _stage_groups(self, stages, domain, order):
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/models/mgmtsystem_nonconformity.py]: Is there any reason for this removal?

File: mgmtsystem_nonconformity/reports/mgmtsystem_nonconformity_report.xml [hunk 8]
```diff
@@ -0,0 +1,47 @@
+<odoo>
+    <!-- Management System Non Conformity Report Pivot View -->
+    <record id="view_report_mgmtsystem_nonconformity_pivot" model="ir.ui.view">
+        <field name="name">mgmtsystem.nonconformity.report.pivot</field>
+        <field name="model">mgmtsystem.nonconformity</field>
+        <field name="arch" type="xml">
+            <pivot string="Non conformity" disable_linking="True">
+                <field name="state" type="row" />
+                <field name="user_id" type="col" />
+                <field name="number_of_nonconformities" type="measure" />
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/reports/mgmtsystem_nonconformity_report.xml:10]:
```suggestion
                <field name="number_of_nonconformities" type="measure" readonly="True"  />
```

File: mgmtsystem_nonconformity/reports/mgmtsystem_nonconformity_report.xml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    <!-- Non Conformity Report Graph View -->
+    <record id="view_report_mgmtsystem_nonconformity_graph" model="ir.ui.view">
+        <field name="name">mgmtsystem.nonconformity.report.graph</field>
+        <field name="model">mgmtsystem.nonconformity</field>
+        <field name="arch" type="xml">
+            <graph string="Non conformity">
+                <field name="state" />
+                <field name="user_id" />
+                <field name="number_of_nonconformities" type="measure" />
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/reports/mgmtsystem_nonconformity_report.xml:23]:
```suggestion
                <field name="number_of_nonconformities" type="measure" readonly="True"  />
```

File: mgmtsystem_nonconformity/views/mgmtsystem_nonconformity.xml [hunk 10]
```diff
@@ -8,11 +8,11 @@
...
 
     <record id="view_mgmtsystem_nonconformity_tree" model="ir.ui.view">
-        <field name="name">mgmtsystem.nonconformity.tree</field>
+        <field name="name">mgmtsystem.nonconformity.list</field>
         <field name="model">mgmtsystem.nonconformity</field>
         <field name="arch" type="xml">
-            <tree>
-                <field name="ref" />
+            <list>
+                <field name="ref" readonly="1" />
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/views/mgmtsystem_nonconformity.xml]:
```suggestion
                <field name="ref" readonly="True" />
```

File: mgmtsystem_nonconformity/models/mgmtsystem_nonconformity.py (hunk 7 above)
Comment (natuan9) [reply mgmtsystem_nonconformity/models/mgmtsystem_nonconformity.py]: order parameter of group_expand methods is remove
you can see the detail here `https://github.com/odoo/odoo/pull/139294/commits/6e6bcbc6ab5cef679b17b617d821b2bb3f778af8`

File: mgmtsystem_nonconformity/models/mgmtsystem_nonconformity_cause.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+    _name = "mgmtsystem.nonconformity.cause"
+    _description = "Cause of the nonconformity of the management system"
+    _order = "parent_id, sequence"
+    _parent_store = True
+
+    name = fields.Char("Cause", required=True, translate=True)
+    description = fields.Text()
+    sequence = fields.Integer(help="Defines the order to present items")
+    parent_path = fields.Char(index=True, unaccent=False)
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/models/mgmtsystem_nonconformity_cause.py]: Remove **`unaccent`** in all fields. It is not necessary anymore.

File: mgmtsystem_nonconformity/views/mgmtsystem_origin.xml [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+    <!-- Origin -->
+
+    <record id="view_mgmtsystem_nonconformity_origin_form" model="ir.ui.view">
+        <field name="name">mgmtsystem.nonconformity.origin.form</field>
+        <field name="type">form</field>
+        <field name="model">mgmtsystem.nonconformity.origin</field>
+        <field name="arch" type="xml">
+            <form string="Origins of nonconformity">
+                <sheet>
+                    <field name="active" invisible="1" />
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/views/mgmtsystem_origin.xml]:
```suggestion

```

File: mgmtsystem_nonconformity/views/mgmtsystem_nonconformity.xml [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+        <field name="model">mgmtsystem.nonconformity</field>
+        <field name="arch" type="xml">
+            <form string="Non-Conformity">
+                <header>
+                    <field
+                        name="stage_id"
+                        widget="statusbar"
+                        options="{'clickable': 1}"
+                    />
+                    <field name="state" invisible="True" />
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/views/mgmtsystem_nonconformity.xml]:
```suggestion

```

File: mgmtsystem_nonconformity/models/mgmtsystem_nonconformity_origin.py [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+class MgmtsystemNonconformityOrigin(models.Model):
+    _name = "mgmtsystem.nonconformity.origin"
+    _description = "Origin of nonconformity of the management system"
+    _order = "parent_id, sequence"
+    _parent_store = True
+
+    name = fields.Char("Origin", required=True, translate=True)
+    description = fields.Text()
+    sequence = fields.Integer(help="Defines the order to present items")
+    parent_path = fields.Char(index=True, unaccent=False)
```
Comment (thienvh332) [inline mgmtsystem_nonconformity/models/mgmtsystem_nonconformity_origin.py]: Remove here

**Review by thienvh332 — APPROVED** [review]
LGTM

**Pull request closed without merging**